File: apps/api/src/services/template.service.ts
import type { HydratedDocument } from "mongoose";
import { interpolateTemplate, type CreateTemplateInput, type Template, type UpdateTemplateInput } from "@believe-ai/shared";
import sanitizeHtml from "sanitize-html";
import type { TemplateDocument } from "@believe-ai/server";
import { templateRepository } from "../repositories/template.repository.js";
import { NotFoundError } from "../errors/AppError.js";

function toDto(doc: HydratedDocument<TemplateDocument>): Template {
  return {
    id: doc._id.toString(),
    userId: doc.userId.toString(),
    name: doc.name,
    subject: doc.subject,
    body: doc.body,
    createdAt: doc.createdAt.toISOString(),
    updatedAt: doc.updatedAt.toISOString(),
  };
}

/** Templates and AI output are treated as untrusted content — strip anything beyond basic formatting. */
function sanitizeBody(body: string): string {
  return sanitizeHtml(body, {
    allowedTags: ["p", "br", "b", "strong", "i", "em", "u", "a", "ul", "ol", "li", "span", "div"],
    allowedAttributes: { a: ["href", "target", "rel"] },
  });
}

export const templateService = {
  toDto,

  async list(userId: string): Promise<Template[]> {
    const docs = await templateRepository.list(userId);
    return docs.map(toDto);
  },

  async getById(id: string, userId: string): Promise<Template> {
    const doc = await templateRepository.findById(id, userId);
    if (!doc) throw new NotFoundError("Template not found");
    return toDto(doc);
  },

  async create(userId: string, input: CreateTemplateInput): Promise<Template> {
    const doc = await templateRepository.create({
      userId,
      name: input.name,
      subject: input.subject,
      body: sanitizeBody(input.body),
    });
    return toDto(doc);
  },

  async update(id: string, userId: string, input: UpdateTemplateInput): Promise<Template> {
    const updates = { ...input, body: input.body ? sanitizeBody(input.body) : undefined };
    const doc = await templateRepository.update(id, userId, updates);
    if (!doc) throw new NotFoundError("Template not found");
    return toDto(doc);
  },

  async delete(id: string, userId: string): Promise<void> {
    const deleted = await templateRepository.delete(id, userId);
    if (!deleted) throw new NotFoundError("Template not found");
  },

  async duplicate(id: string, userId: string): Promise<Template> {
    const original = await templateRepository.findById(id, userId);
    if (!original) throw new NotFoundError("Template not found");
    const copy = await templateRepository.create({
      userId,
      name: `${original.name} (copy)`,
      subject: original.subject,
      body: original.body,
    });
    return toDto(copy);
  },

  preview(subject: string, body: string, values: Record<string, string>) {
    return {
      subject: interpolateTemplate(subject, values),
      body: interpolateTemplate(body, values),
    };
  },
};

File: apps/backend/services/template_service.py
"""Templates — mirrors apps/api's template.service.ts, including HTML
sanitization on write (templates and AI output are untrusted content) and
{{variable}} interpolation for preview."""

from __future__ import annotations

from typing import Any

from bson import ObjectId

from core.errors import NotFoundError
from models.template import Template, TemplateBodyFormat
from repositories import template_repository
from schemas.template import CreateTemplateInput, TemplateDto, UpdateTemplateInput
from services.email_content import interpolate_html, to_html, to_plain_text
from services.personalization import LINK_KEYS, VARIABLE_PATTERN, canonical_key


def _to_dto(doc: Template) -> TemplateDto:
    rendered_html = to_html(doc.body, doc.bodyFormat)
    return TemplateDto(
        id=str(doc.id),
        userId=str(doc.userId),
        name=doc.name,
        subject=doc.subject,
        body=doc.body,
        bodyFormat=doc.bodyFormat,
        bodyHtml=rendered_html,
        bodyText=to_plain_text(rendered_html),
        createdAt=doc.createdAt.isoformat(),
        updatedAt=doc.updatedAt.isoformat(),
    )


def interpolate(text: str, values: dict[str, str]) -> str:
    """Plain (non-HTML-escaping) {{variable}} substitution — used for the
    subject line, which is never HTML, and as the legacy text-body path.
    Unknown or missing variables are left blank rather than raising —
    templates must still render for contacts with partial data.

    Names are folded through the registry's aliases first (see
    services/personalization.py), so {{sender_name}} and {{senderName}}
    resolve to the same profile field."""
    return VARIABLE_PATTERN.sub(lambda m: values.get(canonical_key(m.group(1)), ""), text)


async def list_templates(user_id: ObjectId) -> list[TemplateDto]:
    docs = await template_repository.list_for_user(user_id)
    return [_to_dto(doc) for doc in docs]


async def get_by_id(template_id: ObjectId, user_id: ObjectId) -> TemplateDto:
    doc = await template_repository.find_by_id(template_id, user_id)
    if not doc:
        raise NotFoundError("Template not found")
    return _to_dto(doc)


async def create(user_id: ObjectId, input_: CreateTemplateInput) -> TemplateDto:
    body = to_html(input_.body, input_.bodyFormat) if input_.bodyFormat == "html" else input_.body
    doc = await template_repository.create(user_id, name=input_.name, subject=input_.subject, body=body, body_format=input_.bodyFormat)
    return _to_dto(doc)


async def update(template_id: ObjectId, user_id: ObjectId, input_: UpdateTemplateInput) -> TemplateDto:
    updates: dict[str, Any] = input_.model_dump(exclude_unset=True)
    if updates.get("body") is not None:
        existing = await template_repository.find_by_id(template_id, user_id)
        if not existing:
            raise NotFoundError("Template not found")
        body_format = updates.get("bodyFormat", existing.bodyFormat)
        if body_format == "html":
            updates["body"] = to_html(updates["body"], "html")
    doc = await template_repository.update(template_id, user_id, updates)
    if not doc:
        raise NotFoundError("Template not found")
    return _to_dto(doc)


async def delete(template_id: ObjectId, user_id: ObjectId) -> None:
    deleted = await template_repository.delete(template_id, user_id)
    if not deleted:
        raise NotFoundError("Template not found")


async def duplicate(template_id: ObjectId, user_id: ObjectId) -> TemplateDto:
    original = await template_repository.find_by_id(template_id, user_id)
    if not original:
        raise NotFoundError("Template not found")
    copy = await template_repository.create(
        user_id, name=f"{original.name} (copy)", subject=original.subject, body=original.body, body_format=original.bodyFormat
    )
    return _to_dto(copy)


def preview(subject: str, body: str, body_format: TemplateBodyFormat, values: dict[str, str]) -> tuple[str, str]:
    """Renders exactly what the real send pipeline renders — the composer's
    Preview mode and the campaign recipient preview both call this, so
    neither can show something different from what actually gets sent (see
    services/email_content.py's module docstring for why that mismatch used
    to be the whole bug).

    `values` must already be resolved: the route merges the caller's chosen
    preview recipient with sender values built server-side from their own
    profile, so a preview can neither show a stale sender name nor be told
    to render someone else's identity."""
    rendered_html = to_html(body, body_format)
    # LINK_KEYS carry real <a> markup built by personalization.py — the same
    # raw_keys the send job passes, so a sign-off previews as a live link
    # rather than escaped tag text.
    return interpolate(subject, values), interpolate_html(rendered_html, values, raw_keys=LINK_KEYS)
